woocommerce/woocommerce-ios
# [IAP] Basic UI flow to allow upgrades purchase

Almost half of the 500+ lines is due to renaming. `SubscriptionsView`, and `SubscriptionsViewModelTests` is existing code.

## Description:
This PR tackles creating an initial UI flow to allow merchants to upgrade their plan via an Upgrades View. We do so by splitting the existing responsibility between Upgrades (what plan upgrades a merchant might be entitled to upgrade to) and Subscriptions (what plan the merchant's site currently has).

Along with this, we create a new Upgrades View that will appear when we tap on the "Upgrade Now" CTA banner for Free Trial sites. This Upgrades View will currently show all the available debug plans, as well as the details of Essential Monthly, as is the first plan we intend to use with M1.

<img width="320" alt="Screenshot 2023-06-07 at 14 39 20" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/194ff192-d178-493a-8039-cd3aae494843">


## Changes:
- Renamed existing `UpgradesView` and `UpgradesViewModel` to `SubscriptionsView` and `SubscriptionsViewModel`, to reflect the actual usage of this view as the existing subscriptions management view.
- Created a new `UpgradesView` and `UpgradesViewModel` which deals with upgrading and purchasing a new plan. This view is driven by two view models, as we need information from the current site subscription details, and the potential available upgrades.

## Testing instructions
* With a Free Trial site (can be created via https://woocommerce.com/express/ )
* On a physical device, go to My Store > Tap on "Upgrade Now" banner > See the Upgrades View
* See the Plans view with dynamically loaded information
* The button loading indicator will spin until we can fetch the available products, then we can see "Purchase Debug Monthly"
* Once `Purchase Debug Monthly` button appears, tap on purchase and perform the purchase with your sandbox credentials and wait until the "All is set message" appears.

At this point the upgrade is given, but there's a known issue where we won't see these unless we relaunch the app. By relaunching the app you will see the site is under the "Essential" plan upgrade, and  going to Menu > Subscriptions > you can see the subscription details of 1 month of Essential plan.

## Screenshots

| Purchase flow 1 | Purchase flow 2 | Purchase flow 3 | 
|--------|--------|--------|
| <img width="443" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/75e2fefe-8234-4abe-91a4-9c32fa4cf45e"> | <img width="443" alt="Screenshot 2023-06-06 at 17 06 47" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/89370d84-16e2-4922-abd8-328ac9532887"> | <img width="432" alt="Screenshot 2023-06-06 at 17 07 04" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/97956342-c54a-4412-b9bb-adc25d07c930"> | 


After restarting the app:

| Menu | Subscriptions |
|--------|--------|
![IMG_0076](https://github.com/woocommerce/woocommerce-ios/assets/3812076/42f98c15-4602-45b2-8b7b-4388e2283740) | ![IMG_0077](https://github.com/woocommerce/woocommerce-ios/assets/3812076/3fa4e1fb-5325-45ca-947e-b55770700ac8) |
